SIGSEGV with custom LLB

Using this LLB:

```
{
  "Op": {
    "Op": {
      "Source": {
        "identifier": "docker-image://docker.io/library/ubuntu:18.04"
      }
    }
  },
  "Digest": "sha256:b8902a142f4abc38d506955d881b0d9373fc1b467c4c927560437ac96988f44c",
  "OpMetadata": {
    "caps": null
  }
}
{
  "Op": {
    "inputs": [
      {
        "digest": "sha256:b8902a142f4abc38d506955d881b0d9373fc1b467c4c927560437ac96988f44c",
        "index": 0
      }
    ],
    "Op": {
      "Exec": {
        "meta": {
          "args": [
            "apk",
            "update"
          ]
        }
      }
    }
  },
  "Digest": "sha256:3a5074cab5f9e3045da348b1be7e0f703fc5d14fbb72876fcd907f061559d02f",
  "OpMetadata": {
    "caps": null
  }
}
{
  "Op": {
    "inputs": [
      {
        "digest": "sha256:3a5074cab5f9e3045da348b1be7e0f703fc5d14fbb72876fcd907f061559d02f",
        "index": 0
      }
    ],
    "Op": null
  },
  "Digest": "sha256:f432324de03c32126f0e271ac696bb4fab98db3e6b7e8b442cb375dcef629e44",
  "OpMetadata": {
    "caps": {
      "constraints": true,
      "platform": true
    }
  }
}
```

Buildkitd segfaults at:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x12fc5ab]

goroutine 2242 [running]:
github.com/moby/buildkit/executor/runcexecutor.(*runcExecutor).Exec(0xc00007e7e0, 0x1757900, 0xc0007dc9f0, 0xc000aaa3e0, 0x2, 0x2, 0xc000a19870, 0x1, 0x1, 0x0, ...)
	/home/andy/go/src/github.com/moby/buildkit/executor/runcexecutor/executor.go:150 +0x1eb
github.com/moby/buildkit/solver/llbsolver/ops.(*execOp).Exec(0xc0005ca780, 0x1757900, 0xc0007dc9f0, 0xc000a197b0, 0x1, 0x1, 0x0, 0x0, 0x0, 0x0, ...)
	/home/andy/go/src/github.com/moby/buildkit/solver/llbsolver/ops/exec.go:711 +0x1a6d
github.com/moby/buildkit/solver.(*sharedOp).Exec.func1(0x1757900, 0xc0007dc9f0, 0x0, 0x0, 0x0, 0x0)
	/home/andy/go/src/github.com/moby/buildkit/solver/jobs.go:688 +0x368
github.com/moby/buildkit/util/flightcontrol.(*call).run(0xc000383320)
	/home/andy/go/src/github.com/moby/buildkit/util/flightcontrol/flightcontrol.go:116 +0x65
sync.(*Once).Do(0xc000383378, 0xc000a19800)
	/home/andy/.go/src/sync/once.go:44 +0xb3
created by github.com/moby/buildkit/util/flightcontrol.(*call).wait
	/home/andy/go/src/github.com/moby/buildkit/util/flightcontrol/flightcontrol.go:142 +0x18b
```

The related code is:

```
mountable, err := root.Mount(ctx, false)
```

So I assume `root` is `nil`?  My golang is not super strong, so it's quite possible I'm barking up the wrong tree there.

I guess there are two problems here:

1) What about this LLB means I'm hitting this segfault?  I suspect the `caps: null` is wrong, but I don't know what values make sense there.
2) builtkitd has a "dial-a-boot" where valid LLB causes the daemon to die 💀

--- a/solver/llbsolver/ops/build.go
+++ b/solver/llbsolver/ops/build.go
@@ -10,6 +10,7 @@ import (
 	"github.com/moby/buildkit/frontend"
 	"github.com/moby/buildkit/snapshot"
 	"github.com/moby/buildkit/solver"
+	"github.com/moby/buildkit/solver/llbsolver"
 	"github.com/moby/buildkit/solver/pb"
 	"github.com/moby/buildkit/worker"
 	digest "github.com/opencontainers/go-digest"
@@ -25,6 +26,9 @@ type buildOp struct {
 }
 
 func NewBuildOp(v solver.Vertex, op *pb.Op_Build, b frontend.FrontendLLBBridge, _ worker.Worker) (solver.Op, error) {
+	if err := llbsolver.ValidateOp(&pb.Op{Op: op}); err != nil {
+		return nil, err
+	}
 	return &buildOp{
 		op: op.Build,
 		b:  b,

--- a/solver/llbsolver/ops/exec.go
+++ b/solver/llbsolver/ops/exec.go
@@ -60,6 +60,9 @@ type execOp struct {
 }
 
 func NewExecOp(v solver.Vertex, op *pb.Op_Exec, platform *pb.Platform, cm cache.Manager, sm *session.Manager, md *metadata.Store, exec executor.Executor, w worker.Worker) (solver.Op, error) {
+	if err := llbsolver.ValidateOp(&pb.Op{Op: op}); err != nil {
+		return nil, err
+	}
 	return &execOp{
 		op:          op.Exec,
 		cm:          cm,

--- a/solver/llbsolver/ops/file.go
+++ b/solver/llbsolver/ops/file.go
@@ -35,6 +35,9 @@ type fileOp struct {
 }
 
 func NewFileOp(v solver.Vertex, op *pb.Op_File, cm cache.Manager, md *metadata.Store, w worker.Worker) (solver.Op, error) {
+	if err := llbsolver.ValidateOp(&pb.Op{Op: op}); err != nil {
+		return nil, err
+	}
 	return &fileOp{
 		op:        op.File,
 		md:        md,

--- a/solver/llbsolver/ops/source.go
+++ b/solver/llbsolver/ops/source.go
@@ -7,6 +7,7 @@ import (
 
 	"github.com/moby/buildkit/session"
 	"github.com/moby/buildkit/solver"
+	"github.com/moby/buildkit/solver/llbsolver"
 	"github.com/moby/buildkit/solver/pb"
 	"github.com/moby/buildkit/source"
 	"github.com/moby/buildkit/worker"
@@ -26,6 +27,9 @@ type sourceOp struct {
 }
 
 func NewSourceOp(_ solver.Vertex, op *pb.Op_Source, platform *pb.Platform, sm *source.Manager, sessM *session.Manager, w worker.Worker) (solver.Op, error) {
+	if err := llbsolver.ValidateOp(&pb.Op{Op: op}); err != nil {
+		return nil, err
+	}
 	return &sourceOp{
 		op:       op,
 		sm:       sm,

--- a/solver/llbsolver/vertex.go
+++ b/solver/llbsolver/vertex.go
@@ -188,8 +188,15 @@ func loadLLB(def *pb.Definition, fn func(digest.Digest, *pb.Op, func(digest.Dige
 		allOps[dgst] = &op
 	}
 
+	if len(allOps) < 2 {
+		return solver.Edge{}, errors.Errorf("invalid LLB with %d vertexes", len(allOps))
+	}
+
 	lastOp := allOps[dgst]
 	delete(allOps, dgst)
+	if len(lastOp.Inputs) == 0 {
+		return solver.Edge{}, errors.Errorf("invalid LLB with no inputs on last vertex")
+	}
 	dgst = lastOp.Inputs[0].Digest
 
 	cache := make(map[digest.Digest]solver.Vertex)
@@ -203,6 +210,11 @@ func loadLLB(def *pb.Definition, fn func(digest.Digest, *pb.Op, func(digest.Dige
 		if !ok {
 			return nil, errors.Errorf("invalid missing input digest %s", dgst)
 		}
+
+		if err := ValidateOp(op); err != nil {
+			return nil, err
+		}
+
 		v, err := fn(dgst, op, rec)
 		if err != nil {
 			return nil, err
@@ -238,6 +250,55 @@ func llbOpName(op *pb.Op) string {
 	default:
 		return "unknown"
 	}
+}
+
+func ValidateOp(op *pb.Op) error {
+	if op == nil {
+		return errors.Errorf("invalid nil op")
+	}
+
+	switch op := op.Op.(type) {
+	case *pb.Op_Source:
+		if op.Source == nil {
+			return errors.Errorf("invalid nil source op")
+		}
+	case *pb.Op_Exec:
+		if op.Exec == nil {
+			return errors.Errorf("invalid nil exec op")
+		}
+		if op.Exec.Meta == nil {
+			return errors.Errorf("invalid exec op with no meta")
+		}
+		if len(op.Exec.Meta.Args) == 0 {
+			return errors.Errorf("invalid exec op with no args")
+		}
+		if len(op.Exec.Mounts) == 0 {
+			return errors.Errorf("invalid exec op with no mounts")
+		}
+
+		isRoot := false
+		for _, m := range op.Exec.Mounts {
+			if m.Dest == pb.RootMount {
+				isRoot = true
+				break
+			}
+		}
+		if !isRoot {
+			return errors.Errorf("invalid exec op with no rootfs")
+		}
+	case *pb.Op_File:
+		if op.File == nil {
+			return errors.Errorf("invalid nil file op")
+		}
+		if len(op.File.Actions) == 0 {
+			return errors.Errorf("invalid file op with no actions")
+		}
+	case *pb.Op_Build:
+		if op.Build == nil {
+			return errors.Errorf("invalid nil build op")
+		}
+	}
+	return nil
 }
 
 func fileOpName(actions []*pb.FileAction) string {

